File: docs/blogs/environment/centos/centos7/projects/apollo2.0.1-built-in-eureka/apollo-portal/src/main/java/com/ctrip/framework/apollo/openapi/v1/controller/ClusterController.java
/*
 * Copyright 2022 Apollo Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.ctrip.framework.apollo.openapi.v1.controller;

import com.ctrip.framework.apollo.openapi.api.ClusterOpenApiService;
import com.ctrip.framework.apollo.portal.spi.UserService;
import java.util.Objects;
import javax.servlet.http.HttpServletRequest;
import javax.validation.Valid;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import com.ctrip.framework.apollo.common.exception.BadRequestException;
import com.ctrip.framework.apollo.common.utils.InputValidator;
import com.ctrip.framework.apollo.common.utils.RequestPrecondition;
import com.ctrip.framework.apollo.core.utils.StringUtils;
import com.ctrip.framework.apollo.openapi.dto.OpenClusterDTO;

@RestController("openapiClusterController")
@RequestMapping("/openapi/v1/envs/{env}")
public class ClusterController {

  private final UserService userService;
  private final ClusterOpenApiService clusterOpenApiService;

  public ClusterController(
      UserService userService,
      ClusterOpenApiService clusterOpenApiService) {
    this.userService = userService;
    this.clusterOpenApiService = clusterOpenApiService;
  }

  @GetMapping(value = "apps/{appId}/clusters/{clusterName:.+}")
  public OpenClusterDTO getCluster(@PathVariable("appId") String appId, @PathVariable String env,
      @PathVariable("clusterName") String clusterName) {
    return this.clusterOpenApiService.getCluster(appId, env, clusterName);
  }

  @PreAuthorize(value = "@consumerPermissionValidator.hasCreateClusterPermission(#request, #appId)")
  @PostMapping(value = "apps/{appId}/clusters")
  public OpenClusterDTO createCluster(@PathVariable String appId, @PathVariable String env,
      @Valid @RequestBody OpenClusterDTO cluster, HttpServletRequest request) {

    if (!Objects.equals(appId, cluster.getAppId())) {
      throw new BadRequestException(String.format(
          "AppId not equal. AppId in path = %s, AppId in payload = %s", appId, cluster.getAppId()));
    }

    String clusterName = cluster.getName();
    String operator = cluster.getDataChangeCreatedBy();

    RequestPrecondition.checkArguments(!StringUtils.isContainEmpty(clusterName, operator),
        "name and dataChangeCreatedBy should not be null or empty");

    if (!InputValidator.isValidClusterNamespace(clusterName)) {
      throw new BadRequestException(
          String.format("Invalid ClusterName format: %s", InputValidator.INVALID_CLUSTER_NAMESPACE_MESSAGE));
    }

    if (userService.findByUserId(operator) == null) {
      throw new BadRequestException("User " + operator + " doesn't exist!");
    }

    return this.clusterOpenApiService.createCluster(env, cluster);
  }

}

File: docs/blogs/environment/centos/centos7/projects/apollo2.0.1-built-in-eureka/apollo-portal/src/main/resources/static/scripts/controller/ClusterController.js
/*
 * Copyright 2022 Apollo Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
cluster_module.controller('ClusterController',
    ['$scope', '$location', '$window', '$translate', 'toastr', 'AppService', 'EnvService', 'ClusterService',
        'AppUtil',
        function ($scope, $location, $window, $translate, toastr, AppService, EnvService, ClusterService,
            AppUtil) {

            var params = AppUtil.parseParams($location.$$url);
            $scope.appId = params.appid;

            $scope.step = 1;

            $scope.submitBtnDisabled = false;

            EnvService.find_all_envs().then(function (result) {
                $scope.envs = [];
                result.forEach(function (env) {
                    $scope.envs.push({ name: env, checked: false });

                });
                $(".apollo-container").removeClass("hidden");
            }, function (result) {
                toastr.error(AppUtil.errorMsg(result), $translate.instant('Cluster.LoadingEnvironmentError'));
            });

            $scope.clusterName = '';

            $scope.switchChecked = function (env, $event) {
                env.checked = !env.checked;
                $event.stopPropagation();
            };

            $scope.toggleEnvCheckedStatus = function (env) {
                env.checked = !env.checked;
            };

            $scope.create = function () {

                var noEnvChecked = true;
                $scope.envs.forEach(function (env) {
                    if (env.checked) {
                        noEnvChecked = false;
                        $scope.submitBtnDisabled = true;
                        ClusterService.create_cluster($scope.appId, env.name,
                            {
                                name: $scope.clusterName,
                                appId: $scope.appId
                            }).then(function (result) {
                                toastr.success(env.name, $translate.instant('Cluster.ClusterCreated'));
                                $scope.step = 2;
                                $scope.submitBtnDisabled = false;
                            }, function (result) {
                                toastr.error(AppUtil.errorMsg(result), $translate.instant('Cluster.ClusterCreateFailed'));
                                $scope.submitBtnDisabled = false;
                            })
                    }
                });

                if (noEnvChecked) {
                    toastr.warning($translate.instant('Cluster.PleaseChooseEnvironment'));
                }

            };

        }]);
